tool/gravity/cli/ops.go

```diff
@@ -81,33 +82,87 @@ func uploadUpdate(ctx context.Context, tarballEnv *localenv.TarballEnvironment,
 		return trace.Wrap(err)
 	}
 
-	appPackage, err := install.GetAppPackage(tarballEnv.Apps)
+	application, err := install.GetApp(tarballEnv.Apps)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	env.PrintStep("Importing application %v v%v", appPackage.Name, appPackage.Version)
-	_, err = appservice.PullApp(appservice.AppPullRequest{
-		SrcPack: tarballEnv.Packages,
-		SrcApp:  tarballEnv.Apps,
-		DstPack: clusterPackages,
-		DstApp:  clusterApps,
-		Package: *appPackage,
+	var registries []string
+	err = utils.Retry(defaults.RetryInterval, defaults.RetryLessAttempts, func() error {
+		registries, err = getRegistries(ctx, env, cluster.ClusterState.Servers)
+		return trace.Wrap(err)
 	})
 	if err != nil {
-		if !trace.IsAlreadyExists(err) {
+		return trace.Wrap(err)
+	}
+	if len(registries) == 0 {
+		return trace.NotFound("could not determine cluster registry addresses")
```

Comment:
this sounds like something getRestries should handle - i.e. return a corresponding error for? In this case, the error context might be closer to the error.

Reply:
Not sure I fully follow the proposal here. I still need to check the returned list length to be sure that I can take the first element?